types/state_aware_marker.go

```diff
@@ -0,0 +1,154 @@
+type FushableMarker interface {
+	marker
+	Flush() error
+}
+
+type StateAppender interface {
+	Append(fingerprint model.Fingerprint, state AlertState)
+	AppendInhibited(fingerprint model.Fingerprint, inhibitedBy []string)
+	Close() error
+	Flush() error
+}
+
+func NewStateAwareMarker(
+	r prometheus.Registerer,
+	appender StateAppender,
+) *StateAwareMarker {
+	return &StateAwareMarker{
+		marker:   NewMarker(r),
+		appender: appender,
+	}
+}
+
+type StateAwareMarker struct {
+	marker   marker
+	appender StateAppender
+}
+
+func (s *StateAwareMarker) SetActiveOrSilenced(
+	alert model.Fingerprint,
+	version int,
+	activeSilenceIDs,
+	pendingSilenceIDs []string,
+) {
+	currStatus := s.marker.Status(alert)
+
+	s.marker.SetActiveOrSilenced(alert, version, activeSilenceIDs, pendingSilenceIDs)
+
+	// There is a chance that marker marks it as active, so we need to check.
+	if _, _, _, isSilenced := s.marker.Silenced(alert); isSilenced {
+		if currStatus.State != AlertStateSuppressed {
+			s.appender.Append(alert, AlertStateSuppressed)
+		}
+
+		return
+	}
+
+	if currStatus.State == AlertStateSuppressed && len(currStatus.InhibitedBy) > 0 {
+		return
+	}
+
+	if currStatus.State == AlertStateActive {
+		return
+	}
+
+	s.appender.Append(alert, AlertStateActive)
+}
+
+func (s *StateAwareMarker) SetInhibited(alert model.Fingerprint, alertIDs ...string) {
+	currStatus := s.marker.Status(alert)
+
+	s.marker.SetInhibited(alert, alertIDs...)
+
+	if by, isInhibited := s.marker.Inhibited(alert); isInhibited {
+		if currStatus.State != AlertStateSuppressed {
+			s.appender.AppendInhibited(alert, by)
+
+			return
+		}
+
+		if !slices.Equal(currStatus.InhibitedBy, alertIDs) {
+			s.appender.AppendInhibited(alert, by)
+		}
+
+		return
+	}
+
+	if currStatus.State == AlertStateSuppressed || currStatus.State == AlertStateActive {
+		return
+	}
+
+	s.appender.Append(alert, AlertStateActive)
+}
+
+func (s *StateAwareMarker) Count(states ...AlertState) int {
+	return s.marker.Count(states...)
+}
+
+func (s *StateAwareMarker) Status(alert model.Fingerprint) AlertStatus {
+	return s.marker.Status(alert)
+}
+
+func (s *StateAwareMarker) Delete(alert model.Fingerprint) {
+	currStatus := s.marker.Status(alert)
+
+	s.marker.Delete(alert)
+
+	if currStatus.State == AlertStateUnprocessed {
+		return
+	}
+
+	s.appender.Append(alert, AlertStateDeleted)
+}
+
+func (s *StateAwareMarker) Unprocessed(alert model.Fingerprint) bool {
+	return s.marker.Unprocessed(alert)
+}
+
+func (s *StateAwareMarker) Active(alert model.Fingerprint) bool {
+	return s.marker.Active(alert)
+}
+
+func (s *StateAwareMarker) Silenced(
+	alert model.Fingerprint,
+) (activeIDs, pendingIDs []string, version int, silenced bool) {
+	return s.marker.Silenced(alert)
+}
+
+func (s *StateAwareMarker) Inhibited(alert model.Fingerprint) ([]string, bool) {
+	return s.marker.Inhibited(alert)
+}
+
+func (s *StateAwareMarker) Muted(routeID, groupKey string) ([]string, bool) {
```

Comment:
`IsMuted()` maybe?

Reply:
What is the `[]string` that's returned?

Reply:
Would break the interface: https://github.com/prometheus/alertmanager/blob/c88eb78142c80a8046204650d02ffc7a71cdceda/types/types.go#L113

Reply:
Maybe a comment that says what it returns then?

Reply:
The methods are documented on the interface, are you suggesting that I copy paste these even though the implementation doesn't do anything extra? 

---

I'll add one big doc comment on the struct to explain however that the only diff between this and MemMarker is the fact that it stores the state changes of each alert 👀